### DISCUSSION TITLE
CreateMany while connecting to many

### DISCUSSION THREAD

**Discussion by MatheusGNachtigall** [prisma/prisma#22452] [Q&A] [answered]
### Question

Assuming I have a schema like this (Simplified for explanation):

```
model Post {
  title      String
  categories Category[]
  id         String     @id @default(uuid())
}

model Category {
  name  String @unique
  posts Post[]
  id    String @id @default(uuid())
}
```
and that I already have a list of categories created in my database (say 50+ categories)

Is there a way that I can do something like this: 
```
const postsData = [
  {
    title: "Post 1",
    categories: {
      connect: [{ name: "Category1" }, { name: "Category2" }],
    },
  },
  {
    title: "Post 2",
    categories: {
      connect: [{ name: "Category2" }, { name: "Category3" }],
    },
  },
  // Add more posts as needed
];

const createdPosts = await prisma.post.createMany({
  data: postsData,
});
```

For context, I am migrating databases with 1M+ records where they have a many-to-many relation with another table (like 1M+ posts connected to 500k+ categories)

I wouldnt mind if I needed to do a 2-step creation (Like creating all posts and then creating the connection on the implicit table), but I could not find any way of doing this without having a loop and doing only a **single post creation** at a time, which seems terrible.

Thanks in advance

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @MatheusGNachtigall 👋 

Thank you for raising this question.

Prisma does not support the createMany operation with nested writes, which includes the connect operation for relations. The createMany operation is designed to create multiple records in a single model without handling relations. See this [section](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#create-multiple-records-and-multiple-related-records) of the documentation.

However, you can use the create operation to create a single record along with its relations in a nested write. For example, to create a `Post` and connect it to existing `Category` records by their unique identifier, you would use the following code:
```js
const createPostWithCategories = await prisma.post.create({
  data: {
    title: 'Post 1',
    categories: {
      connect: [{ id: 'category-id-1' }, { id: 'category-id-2' }],
    },
  },
});
```
In your case, since you want to create multiple posts and connect them to existing categories, you would need to loop through your postsData array and call prisma.post.create for each post. Here's an example of how you might do this:
```
async function createPostsAndConnectCategories(postsData) {
  return Promise.all(
    postsData.map(postData =>
      prisma.post.create({
        data: postData,
      })
    )
  );
}

const postsData = [
  {
    title: "Post 1",
    categories: {
      connect: [{ name: "Category1" }, { name: "Category2" }],
    },
  },
  {
    title: "Post 2",
    categories: {
      connect: [{ name: "Category2" }, { name: "Category3" }],
    },
  },
  // Add more posts as needed
];

const createdPosts = await createPostsAndConnectCategories(postsData);
```
This approach will execute the creation of each post and its connections to categories in parallel, which may help with performance compared to a sequential loop. However, it's important to note that this could still be a time-consuming operation given the large number of records you mentioned.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by MatheusGNachtigall** [reply]
Yeah bummer....I hoped to avoid needing to create individually...seems like it isnt possible.

The parallel creation with promises is a nice tip though...will definetly try it out to see performance diferences.

Thanks!
